src/rust/multi.ninemanga/src/parser.rs

```diff
@@ -0,0 +1,270 @@
+use aidoku::{
+	error::Result, prelude::format, std::net::HttpMethod, std::net::Request, std::String, std::Vec,
+	Chapter, DeepLink, Filter, FilterType, Manga, MangaContentRating, MangaPageResult, MangaStatus,
+	MangaViewer, Page,
+};
+
+use crate::helper::*;
+
+pub struct NineMangaSource {
+	pub base_url: String,
+	pub language: String,
+	pub date_format: &'static str,
+	pub completed_series: &'static str,
+	pub date_locale: &'static str,
+}
+
+impl Default for NineMangaSource {
+	fn default() -> Self {
+		NineMangaSource {
+			base_url: String::from("https://www.ninemanga.com"),
+			language: String::from("en"),
+			date_format: "MMM d, yyyy",
+			completed_series: "Completed Series",
+			date_locale: "en_US",
+		}
+	}
+}
+
+impl NineMangaSource {
+	// parse the homepage and filters
+	pub fn parse_manga_list(&self, filters: Vec<Filter>, page: i32) -> Result<MangaPageResult> {
+		let mut included_tags: Vec<String> = Vec::new();
+		let mut excluded_tags: Vec<String> = Vec::new();
+		let mut title: String = String::new();
+		let mut status: String = String::new();
+		for filter in filters {
+			match filter.kind {
+				FilterType::Title => {
+					title = filter.value.as_string()?.read();
+				}
+				FilterType::Genre => match filter.value.as_int().unwrap_or(-1) {
+					0 => excluded_tags.push(filter.object.get("id").as_string()?.read()),
+					1 => included_tags.push(filter.object.get("id").as_string()?.read()),
+					_ => continue,
+				},
+				FilterType::Select => {
+					if filter.name.as_str() == self.completed_series {
+						match filter.value.as_int().unwrap_or(-1) {
+							1 => status = String::from("yes"),
+							2 => status = String::from("no"),
+							_ => continue,
+						}
+					}
+				}
+				_ => continue,
+			};
+		}
+		let url = get_search_url(
+			self.base_url.clone(),
+			title,
+			included_tags,
+			excluded_tags,
+			status,
+			page,
+		);
+		Self::parse_manga_listing(self, url, page)
+	}
+
+	// parse the listing page (popular, latest , new etc)
+	pub fn parse_manga_listing(&self, url: String, _page: i32) -> Result<MangaPageResult> {
+		let mut mangas: Vec<Manga> = Vec::new();
+		let html = Request::new(url.as_str(), HttpMethod::Get).html();
+		for manga in html.select(".direlist dl").array() {
+			let manga_node = manga.as_node();
+			let title = manga_node.select("a.bookname").text().read();
+			let id = manga_node.select("a.bookname").attr("href").read();
```

Comment:
Do not set a full url as a manga id. Use the path and prepend the base url when you actually need the full url. Using the url as an id is bad practice since it will break if the website changes domains.